packages/iTwinUI-react/src/core/TimePicker/TimePicker.tsx

```diff
@@ -470,4 +653,117 @@ const TimePickerColumn = <T,>(props: TimePickerColumnProps<T>): JSX.Element => {
   );
 };
 
+type VirtualizedTimePickerColumnProps<T = Date> = {
+  /**
+   * What value to display in every cell.
+   */
+  valueRenderer: (value: T, precision: Precision) => React.ReactNode;
+  /**
+   * Precision of the time.
+   * @default 'minutes'
+   */
+  precision: Precision;
+} & ClassNameProps &
+  Omit<TimePickerColumnProps<T>, 'valueRenderer'>;
+
+const VirtualizedTimePickerColumn = <T,>(
+  props: VirtualizedTimePickerColumnProps<T>,
+): JSX.Element => {
+  const {
+    data,
+    onFocusChange,
+    onSelectChange,
+    isSameFocused,
+    isSameSelected,
+    setFocus = false,
+    valueRenderer,
+    precision = 'minutes',
+    className = 'iui-time',
+  } = props;
+
+  const needFocus = React.useRef(setFocus);
+
+  // Used to focus row only when changed (keyboard navigation)
+  // e.g. without this on every rerender it would be focused
+  React.useEffect(() => {
+    if (needFocus.current) {
+      needFocus.current = false;
+    }
+  });
+
+  const virtualizedItemRenderer = React.useCallback(
+    (index: number) => {
+      const value = data[index];
+      const isSameFocus = isSameFocused(value);
+      return (
+        <li
+          onKeyDown={(event) => {
+            handleTimeKeyDown(
+              event,
+              data.length - 1,
+              (index) => onFocusChange(data[index]),
+              (index) => onSelectChange(data[index]),
+              index,
+              needFocus,
+            );
+          }}
+          className={cx({
+            'iui-selected': isSameSelected(value),
+          })}
+          key={index}
+          tabIndex={isSameFocus ? 0 : undefined}
+          onClick={() => {
+            onSelectChange(value);
+          }}
+        >
+          {valueRenderer(value, precision)}
+        </li>
+      );
+    },
+    [
+      data,
+      isSameFocused,
+      isSameSelected,
+      onFocusChange,
+      onSelectChange,
+      precision,
+      valueRenderer,
+    ],
+  );
+
+  const currentDateIndex = () => {
+    for (let i = 0; i < data.length; i++) {
+      if (isSameSelected(data[i])) {
+        return i;
+      }
+    }
+    return -1;
+  };
+
+  const { outerProps, innerProps, visibleChildren } = useVirtualization({
+    itemsLength: data.length,
+    itemRenderer: virtualizedItemRenderer,
+    scrollToIndex: currentDateIndex(),
+  });
+
+  return (
```

Comment:
Is it possible to abstract the return part in some way so that both `TimePickerColumn` and `VirtualizedTimePickerColumn` can use it?

Reply:
I don't think so; they're pretty different: 
![image](https://user-images.githubusercontent.com/37936169/207722612-b5429977-cfac-4685-a72c-7feb607d9196.png)